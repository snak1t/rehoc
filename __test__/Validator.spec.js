import React from 'react'
import { shallow, mount } from 'enzyme'
import withValidation from '../src/Validator'

const DummyForm = ({ login, asyncField }) =>
  <form>
    <input
      type="text"
      className="loginInput"
      value={login.value}
      onChange={login.handler}
    />
    <input
      type="text"
      className="asyncInput"
      value={asyncField.value}
      onChange={asyncField.handler}
    />
  </form>

const msg = 'Name must be John'
const validationConfig = {
  login: {
    validators: [{ rule: name => name === 'John', message: msg }]
  },
  asyncField: {
    validators: [
      {
        rule: (value, done) => setTimeout(() => done(value), 100),
        message: 'AsyncMsg',
        async: true
      }
    ]
  }
}

describe('Validator', () => {
  it('should exist', () => {
    const Component = withValidation(validationConfig)(DummyForm)
    const component = mount(<Component />)
    const testForm = component.find(DummyForm)
    expect(component.exists()).toEqual(true)
    expect(testForm.props()).toMatchSnapshot()
  })

  it('should update a string value', () => {
    const Component = withValidation(validationConfig)(DummyForm)
    const component = mount(<Component />)
    const testForm = component.find(DummyForm)
    const input = component.find('.loginInput')
    input.simulate('change', { target: { value: 'Sam' } })
    expect(testForm.props().login.value).toEqual('Sam')
  })

  it('should mark a field as dirty if changed', () => {
    const Component = withValidation(validationConfig)(DummyForm)
    const component = mount(<Component />)
    const testForm = component.find(DummyForm)
    const input = component.find('.loginInput')
    input.simulate('change', { target: { value: 'Sam' } })
    expect(testForm.props().login.status.dirty).toEqual(true)
  })

  it('should mark a field as valid if provided with valid value', () => {
    const Component = withValidation(validationConfig)(DummyForm)
    const component = mount(<Component />)
    const testForm = component.find(DummyForm)
    const input = component.find('.loginInput')
    input.simulate('change', { target: { value: 'Sam' } })
    expect(testForm.props().login.status.valid).toEqual(false)
    input.simulate('change', { target: { value: 'John' } })
    expect(testForm.props().login.status.valid).toEqual(true)
  })

  it('should provide an error array if specified value is incorrect', () => {
    const Component = withValidation(validationConfig)(DummyForm)
    const component = mount(<Component />)
    const testForm = component.find(DummyForm)
    const input = component.find('.loginInput')
    input.simulate('change', { target: { value: 'Sam' } })
    expect(testForm.props().login.status.valid).toEqual(false)
    expect(testForm.props().login.errors).toBeInstanceOf(Array)
    expect(testForm.props().login.errors.length).toEqual(1)
    expect(testForm.props().login.errors[0]).toEqual(msg)
  })

  it('should also deal with async validations', () => {
    const waiter = () =>
      new Promise((res, rej) => {
        setTimeout(() => res(), 500)
      })
    const Component = withValidation(validationConfig)(DummyForm)
    const component = mount(<Component />)
    const testForm = component.find(DummyForm)
    const input = component.find('.asyncInput')
    input.simulate('change', { target: { value: true } })
    return waiter().then(() => {
      expect(testForm.props().asyncField.status.valid).toEqual(true)
    })
  })
})
