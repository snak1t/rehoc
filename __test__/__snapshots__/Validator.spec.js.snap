// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator should exist 1`] = `
Object {
  "asyncField": Object {
    "errors": Array [],
    "handler": [Function],
    "status": Object {
      "dirty": false,
      "valid": false,
    },
    "value": "",
  },
  "login": Object {
    "errors": Array [],
    "handler": [Function],
    "status": Object {
      "dirty": false,
      "valid": false,
    },
    "value": "",
  },
  "valid": false,
}
`;
